proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,403 @@
+
+## Appendix 1: LLVM Community Metrics Methods
+
+This document references some LLVM community metrics. This appendix contains the methodology and tools used to produce the metrics cited above. The metrics are gathered from git metadata with a scoped date range of September 1, 2022 through September 1, 2023. Using a specific date range allows these metrics to be re-generated consistently, however the metrics should be extended to encompass the latest usage patterns when driving decision making.
+
+**These scripts are not intended to drive voter eligibility.** These scripts computed data that informed the formation of _area teams_, but the data is severely limited and insufficient for application to voter eligibility.
+
+### git-crawler.py
+
+This script traverses the directories in the LLVM project listing all the authors of commits in the directory once per commit. The script sums up the number of times an authors name appears to produce an approximation of the number of unique authors who have contributed code under a specific sub-directory. The script does not traverse into sub-directories that have less than 50 unique authors, and it skips `test` and `unittests` folders.
+
+The output printed is the number of unique authors that contribute to each directory with 50 or more authors, and a list of the top ten authors that committed to that directory with their number of commits.
+
+```python
+#!/usr/bin/env python3
+
+import os
+import subprocess
+from collections import Counter
+
+def handleDirectory(path):
+ path = path.replace(os.getcwd() + '/', ':')
+ output = subprocess.check_output(['git', 'log', '--format=%an', '--since=September 1 2022', '--until=September 1 2023', '--', path]).decode('utf-8').splitlines()
+ authors = {}
+ for line in output:
+   if line not in authors:
+     authors[line] = 0
+   authors[line] += 1
+ return authors
+
+results = {}
+results[os.getcwd()] = handleDirectory(os.getcwd())
+i = 0
+for root, dirs, files in os.walk(os.getcwd()):
+ i += 1
+ if root not in results:
+   dirs = []
+ toRemove = []
+ for dir in dirs:
+   if dir == 'test' or dir == 'unittests':
+     toRemove.append(dir)
+     continue
+
+   path = os.path.join(root, dir)
+   results[path] = handleDirectory(path)
+   if len(results[path]) < 50:
+     toRemove.append(dir)
+   else:
+     print('%s: %d' % (path, len(results[path])) )
+ for dir in toRemove:
+   dirs.remove(dir)
+
+for dir in results:
+ if not results[dir]:
+   continue
+ counters = Counter(results[dir])
+ high = counters.most_common(10)
+ print('%s:' % dir)
+ for val in high:
+   print('  %s: %d' % (str(val[0]), val[1]))
+```
+
+### Quick Sum Commits by Author for a Directory
```

Comment:
By the way, all these tools will hit in their way this issue: https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223?u=boomanaiden154-1

Reply:
Yeah, I don't think there's a good way to do that, and in other discussions there's enough points to make "number of commits", "number of lines" or any other metric totally unsuitable to measure "representation". In fact, there's enough evidence in other projects and democratic groups that any kind of segmentation leads to bias and mis-representation.

I think we need to take a step back and think hard about what we're trying to accomplish. This document has a lot of unproven proposed solutions for hypothetical problems that we don't even know we'll have. This has never been our community approach and I don't think we need this now.

I propose we clean it up, make it super simple, use some tools to register people (SSO Github) and then see what happens. None of our policies are binding in a way that can't be reverted, so we can stop a derailing voting before it makes damage. We also already have the power to stop misbehavior (CoC), 

I don't see why we're trying so hard to encode hypothetical possibilities. Premature optimization?

Reply:
These tools were not intended to drive voter eligibility. I just wrote them to help gather some (admittedly poor quality) data to help inform choosing _area teams_. I think to drive actual voting we'll need a voter registration system where voters provide personally identifying information that we can use to map to commits, forum posts, code review activity, and anything else we can script.

Reply:
The proposal isn't intended to change how we nominate or approve code owners. I'll update the document to spell this out more clearly, but the current process of posting to Discourse and waiting to see if anyone objects will be maintained. As with the current policy any individual can nominate any other individual to be a code owner.

The only thing this proposal changes is making someone responsible for maintaining the code owners list. That mostly amounts to giving someone the clear authority to remove stale owners and go seek out new ones.